File: Checks/PWR009/example-omp.c
// PWR009: Use OpenMP teams to offload work to GPU

void example(double (*A)[100], double (*B)[100], double (*C)[100]) {
  #pragma omp target map(to: A[0:100][0:100], B[0:100][0:100]) \
          map(tofrom: C[0:100][0:100])
  {
    #pragma omp parallel default(none) shared(A, B, C)
    {
      #pragma omp for schedule(auto)
      for (int i = 0; i < 100; i++) {
        for (int j = 0; j < 100; j++) {
            for (int k = 0; k < 100; k++) {
              C[i][j] += A[i][k] * B[k][j];
            }
        }
      }
    } // end parallel
  } // end target
}

File: Checks/PWR009/example-omp.f90
! PWR009: Use OpenMP teams to offload work to GPU

subroutine example(A, B, C)
  implicit none
  real, dimension(:, :), intent(in) :: A, B
  real, dimension(:, :), intent(inout) :: C
  integer :: i, j, k

  !$omp target map(to: A, B) map(tofrom: C)
  !$omp parallel default(none) private(i, j, k) shared(A, B, C)
  !$omp do
  do j = 1, size(C, 2)
    do k = 1, size(C, 2)
      do i = 1, size(C, 1)
        C(i, j) = C(i, j) + A(i, k) * B(k, j)
      end do
    end do
  end do
  !$omp end do
  !$omp end parallel
  !$omp end target
end subroutine example
